Use antsibull-changelog instead of packaged changelog generator

felixfontein opened the pull request:
##### SUMMARY
This is a WIP to replace the packaged [changelog generator](https://github.com/ansible/ansible/blob/devel/packaging/release/changelogs/changelog.py) with https://github.com/felixfontein/ansible-changelog/, a extended version which also works for collections. The config is adjusted to output `changelog.yaml` in the same format as ACD expects ([format](https://github.com/felixfontein/ansible-changelog/blob/master/CHANGELOG-FORMAT.md), [example](https://github.com/felixfontein/ansible/blob/changelog-2.10/changelogs/changelog.yaml)) from collections, instead of the previous `.changes.yaml` file (see [example from stable-2.9](https://github.com/ansible/ansible/blob/stable-2.9/changelogs/.changes.yaml)). ([Example of generated .rst file](https://github.com/felixfontein/ansible/blob/changelog-2.10/changelogs/CHANGELOG-v2.10.rst).)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
changelog

felixfontein committed:
Replace included changelog tool with ansible-changelog.

felixfontein commented:
CC @relrod

felixfontein commented:
CC @danielmellado

felixfontein commented:
The current problem is caused by `pip` executing itself while putting `/usr/local/lib/python3.6/dist-packages` into the path, which causes `enum34`'s `enum` package to be used instead of the built-in `enum` package. That in turn breaks `re`. @abadger is trying to get this fixed in pip.

I'm not sure why `enum34` is installed in the default container; it is not an explicit requirement (https://github.com/ansible/default-test-container/tree/master/requirements), but is dragged along by some dependency (https://github.com/ansible/default-test-container/blob/master/freeze/3.6.txt#L64).

abadger commented:
Here's the uptream pip bug.

abadger commented:
I'm not hopeful that the upstream bug will be fixed soon (It seems like upstream wants to figure out a more invasive fix than what I proposed) so @mattclay, if it turns out that enum34 doesn't need to be installed for python3, that might be the faster way to fix this.

felixfontein commented:
/rebuild

felixfontein commented:
/rebuild_failed

felixfontein commented:
ready_for_review

@bcoca with this PR #69454 will also run better because you don't need to update the changelog generator anymore :)

mattclay commented on test/sanity/requirements.txt:
```diff
@@ -2,3 +2,6 @@ packaging  # needed for update-bundled and changelog
 straight.plugin ; python_version >= '3.5' # needed for hacking/build-ansible.py which will host changelog generation and requires python 3+
+
+# changelog build requires python 3.6+
+antsibull ; python_version >= '3.6'
```
We don't yet have a way to limit code-smell tests to specific python versions. We'll need to add that so the test can be skipped on Python 3.5.

felixfontein replied:
I've implemented `minimum_python_version` in 110e542784b5ceb41cd9bf33964cef06859686c9

mattclay commented on test/lib/ansible_test/_internal/sanity/__init__.py:
```diff
@@ -8,6 +8,8 @@
 import re
 import collections
 
+from distutils.version import LooseVersion
```
Use `str_to_version` instead of `LooseVersion`:

https://github.com/ansible/ansible/blob/d7618712fd13804de07dacf06c64e7ac34c395fc/test/lib/ansible_test/_internal/util.py#L861-L863

We don't need more complex version handling for Python versions in this case.

felixfontein replied:
Didn't knew that one existed. Using it now. Thanks!

felixfontein committed:
Also add antsibull-changelog requirement to package-data test requirements.

felixfontein committed:
Fix ancestor version.

felixfontein committed:
Bump dependency to antsibull-changelog==0.2.1.

felixfontein committed:
Force use of ansible-doc for listing plugins.

mattclay commented on test/sanity/code-smell/changelog.py:
```diff
@@ -18,7 +18,7 @@ def main():
             print('%s:%d:%d: extension must be one of: %s' % (path, 0, 0, ', '.join(allowed_extensions)))
 
-    cmd = ['packaging/release/changelogs/changelog.py', 'lint'] + paths
+    cmd = ['antsibull-changelog', 'lint'] + paths
```
We should invoke the module with `-m` to make sure the correct interpreter is used:

```suggestion
    cmd = [sys.executable, '-m', 'antsibull_changelog', 'lint'] + paths
```

However, it looks like that method of invocation is not currently supported by `antsibull-changelog`.

Yes, this was an existing problem with the old script invocation, but now that we're changing it we should fix it.

mattclay replied:
Alternatively, if the module provides a public API we could just import it, which would be even better.

mattclay replied:
Hmm... I guess that doesn't entirely fix the problem either, because `antsibull-changelog` also uses `subprocess` to invoke `ansible-doc` which may pick up the wrong one, or the wrong (or no) python.

felixfontein replied:
https://github.com/ansible-community/antsibull-changelog/pull/12

felixfontein committed:
Run via python -m.

felixfontein committed:
Also fixed the *other* makefile.

mattclay commented:
@felixfontein We should probably promote this sanity test to be included with `ansible-test` so that it can be executed on collections. Doing so would not force collections to use this tool, since it will only run if there are files in the `changelogs/fragments/` directory.

felixfontein committed:
Move changelog code-smell test from ansible/ansible-specific to contained in ansible-test.

felixfontein committed:
Bump antsibull-changelog to 0.3.1.

felixfontein committed:
Make sure changelog test is watching changelogs/config.yaml, and fails if it is not present.
Also ignore fragment files starting with `.`.

felixfontein committed:
Add changelog fragment.

felixfontein committed:
Mention antsibull-changelog and the new categories in docs/docsite/rst/dev_guide/testing/sanity/changelog.rst

The new categories are added by #69968.

felixfontein committed:
Run changelog test once.

felixfontein committed:
Place requirements/constraints correctly.

felixfontein commented:
I moved the test so it is part of `ansible-test sanity` for collections as well, and adjusted it to properly handle changelogs/config.yaml. I've tested it with several collections with changelogs/config.yaml set up, and with several that haven't (there the test is skipped since neither config nor fragments are present).

The test will fail with collections which have changelogs/fragments/ but no changelogs/config.yaml (f.ex. ansible.windows):
```
ERROR: Found 1 changelog issue(s) which need to be resolved:
ERROR: changelogs/config.yaml:0:0: config file does not exist
See documentation for help: https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/changelog.html
```

felixfontein committed:
Remove version from package-data requirements. Constraints will enforce the correct version.

mattclay commented on test/lib/ansible_test/_data/sanity/code-smell/changelog.py:
```diff
@@ -12,13 +12,31 @@ def main():
+    if not paths_to_check:
+        return
+
+    cmd = [sys.executable, '-m', 'antsibull_changelog', 'lint'] + paths_to_check
```
Shouldn't this also be using the `--use-ansible-doc` option?

mattclay replied:
Oh, wait, I see that option isn't available for `lint`.

felixfontein replied:
It's only needed when scanning for plugins.

mattclay replied:
Yeah, I realized that after looking at the code.

felixfontein commented:
@mattclay @relrod thanks a lot for reviewing and merging!
